plugins/dap/default.nix

```diff
@@ -0,0 +1,200 @@
+    };
+  };
+
+  serverAdapterOption = mkAdapterType {
+    host = helpers.defaultNullOpts.mkStr "127.0.0.1" "Host to connect to";
+
+    port = helpers.mkNullOrOption (types.either types.int types.str) ''
+      Port to connect to.
+      If "$\{port}" dap resolves a free port.
+      This is intended to be used with `executable.args`
+    '';
+
+    executable = {
+      command = helpers.mkNullOrOption types.str "Command that spawns the adapter";
+
+      args = helpers.mkNullOrOption (types.listOf types.str) "command arguments";
+
+      detached = helpers.defaultNullOpts.mkBool true "Spawn the debug adapter in detached state";
+
+      cwd = helpers.mkNullOrOption types.str "Working directory";
+    };
+
+    options.maxRetries = helpers.defaultNullOpts.mkInt 14 ''
+      Amount of times the client should attempt to connect before erroring out.
+      There is a 250ms delay between each retry.
+    '';
+  };
+
+  mkAdapterOption = name: type:
+    helpers.mkNullOrOption (with types; attrsOf (either str type)) ''
+      Debug adapters of `${name}` type.
+      The adapters can also be set to a function which takes three arguments:
+
+      - A `on_config` callback. This must be called with the actual adapter table.
+      - The |dap-configuration| which the user wants to use.
+      - An optional parent session. This is only available if the debug-adapter
+        wants to start a child-session via a `startDebugging` request.
+
+      This can be used to defer the resolving of the values to when a configuration
+      is used. A use-case for this is starting an adapter asynchronous.
+    '';
+
+  mkSignOption = default: desc: {
+    text = helpers.defaultNullOpts.mkStr default desc;
+    texthl = helpers.mkNullOrOption types.str "texthl for sign";
+    linehl = helpers.mkNullOrOption types.str "linehl for sign";
+    numhl = helpers.mkNullOrOption types.str "numhl for sign";
+  };
+
+  processAdapters = type: adapters:
+    with builtins;
+      mapAttrs (_: adapter:
+        if typeOf adapter == "string"
+        then helpers.mkRaw adapter
+        else
+          filterAttrs (n: _: n != "enrichConfig") (
+            adapter
+            // {
+              inherit type;
+              enrich_config = helpers.ifNonNull' adapter.enrichConfig (helpers.mkRaw adapter.enrichConfig);
+            }
+          ))
+      adapters;
+in {
+  imports = [
+    ./dap-go.nix
+    ./dap-python.nix
+    ./dap-ui.nix
+    ./dap-virtual-text.nix
+  ];
+
+  options.plugins.dap =
+    helpers.extraOptionsOptions
+    // {
+      enable = mkEnableOption "dap";
+
+      package = helpers.mkPackageOption "dap" pkgs.vimPlugins.nvim-dap;
+
+      adapters = helpers.mkCompositeOption "dap adapters" {
+        executables = mkAdapterOption "executable" executableAdapterOption;
+        servers = mkAdapterOption "server" serverAdapterOption;
+      };
+
+      configurations =
+        helpers.mkNullOrOption (with types; attrsOf (listOf attrs))
```

Comment:
maybe we can use a freeform option, and requiring the `type`, `request` and `name` fields ?